python/rmm/_lib/torch_allocator.pyx

```diff
@@ -0,0 +1,17 @@
+from cuda.ccudart cimport cudaStream_t
+
+from rmm._lib.memory_resource cimport device_memory_resource
+from rmm._lib.per_device_resource cimport get_current_device_resource
+
+
+cdef public void* allocate(
+    ssize_t size, int device, void* stream
+) except * with gil:
+    cdef device_memory_resource* mr = get_current_device_resource()
+    return mr[0].allocate(size, <cudaStream_t> stream)
+
+cdef public void deallocate(
+    void* ptr, ssize_t size, void* stream
+) except * with gil:
+    cdef device_memory_resource* mr = get_current_device_resource()
+    mr[0].deallocate(ptr, size, <cudaStream_t> stream)
```

Comment:
Why are these gil requiring functions? It seems like it's all pure C code here, no Python objects etc.

Reply:
To be clear, if they're getting called from GIL-requiring code in PyTorch, so be it. I just don't see a reason that these functions need to explicitly acquire the GIL. If PyT can call these in a nogil context, is there a reason for us not to allow that?

Reply:
Because the `allocate()` and `deallocate()` methods _can_ involve Python operations on Python objects,  e.g., in `CallbackMemoryResource` or `FailureCallbackResourceAdaptor`.

Reply:
I was thinking that would be automatically be handled when those callbacks are invoked since those callbacks are stored as Python objects. Those are stored as Python objects in the class, so any interaction with them should reacquire the GIL already, right? I guess the potential issue is that we cast these to `void *` pointers before passing them to the C++ classes, so at the point of the call we've lost Cython's safety net. Is that right? If so, we should consider (out of scope for this PR of course) inserting the necessary Python C API calls into the relevant rmm C++ classes i.e. in `failure_callback_resource_adaptor::do_allocate`.

Reply:
My proposal above was wrong, there's no reason to embed this information in librmm where the callbacks could be anything (not necessarily Python objects). However, it should be the responsibility of the callbacks in rmm's Cython code to acquire the GIL as needed, and we do appear to do this correctly already. The `_oom_callback_function` used by the `FailureCallbackResourceAdaptor` [acquires the GIL before calling the user-provided callback](https://github.com/rapidsai/rmm/blob/branch-23.02/python/rmm/_lib/memory_resource.pyx#L820), as do [both the allocate and deallocate callbacks used by the CallbackMemoryResource](https://github.com/rapidsai/rmm/blob/branch-23.02/python/rmm/_lib/memory_resource.pyx#L544).

Reply:
Yup -- the GIL should neither be released in C++, nor can it be released in Python. The Cython "wrapper" functions are what need to take on the responsibility of handling the GIL.

Reply:
Vyas and I did a bit more exploration of exactly _why_ we need a `with gil` here and ended up quite deep in the CPython and Cython internals (still without a clear answer though). 

The _symptom_ though is clear. If you take the example I have in the PR description:

```
import rmm                                                                                                                                                                                                 
import torch

torch.cuda.memory.change_current_allocator(rmm.rmm_torch_allocator)
                                                                                                                                                                                                           
base_mr = rmm.mr.CudaMemoryResource()                                                                                                                                                                      
                                                                                                                                                                                                           
def allocate_func(size):                                                                                                                                                                                   
    print(f"Allocating {size} bytes")                                                                                                                                                                      
    return base_mr.allocate(size)                                                                                                                                                                          
                                                                                                                                                                                                           
def deallocate_func(ptr, size):                                                                                                                                                                            
    print(f"Deallocating {size} bytes")                                                                                                                                                                    
    return base_mr.deallocate(ptr, size)                                                                                                                                                                   
                                                                                                                                                                                                           
rmm.mr.set_current_device_resource(                                                                                                                                                                        
    rmm.mr.CallbackMemoryResource(allocate_func, deallocate_func)                                                                                                                                          
)                                                                                                                                                                                                          
                                                                                                                                                                                                           
x = torch.tensor([1, 2]).cuda()                                                                                                                                                                            
del x                                                                                                                                                                                                      
y = torch.tensor([1, 2, 3]).cuda()                                                                                                                                                                         
del y                     
```

And raise an error in `allocate_func`, while removing the `with gil`, you'll see that the error is uncaught and eventually this segfaults.